### DISCUSSION TITLE
Where does carapace get its completions from?

### DISCUSSION THREAD

**Discussion by mastermach50** [carapace-sh/carapace-bin#2094] [Q&A] [answered]
I have two questions

1. Where does carapace get completions from?
2. If it has to download them from somewhere do I have to be always connected to the internet ?

**Comment by rsteube** [comment, marked as answer]
> 1. Where does carapace get completions from?

They are compiled into the binary. Some like [`gcloud`](https://github.com/rsteube/carapace-bin/blob/bae9485ea95958e9f45490605aa555d17f63435c/completers/gcloud_completer/cmd/root.go) are bridged.
Where appropriate, commands are executed for argument completion.
E.g. `git` is used to list [branches](https://github.com/rsteube/carapace-bin/blob/bae9485ea95958e9f45490605aa555d17f63435c/pkg/actions/tools/git/branch.go#L23C11-L23C11), `gh` for [repositories](https://github.com/rsteube/carapace-bin/blob/bae9485ea95958e9f45490605aa555d17f63435c/pkg/actions/tools/gh/repository.go#L35) and so on.

> 2. If it has to download them from somewhere do I have to be always connected to the internet ?

No, but you will get an error message when trying to complete an argument that tries to access the internet.

**Reply by mastermach50** [reply]
thanks
